src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    internal static class SseIntrinsics
+    {
+        private static unsafe Vector128<float> Load1(float* src, int* idx)
+        {
+            return Sse.SetScalarVector128(src[idx[0]]);
+        }
+
+        private static unsafe Vector128<float> Load4(float* src, int* idx)
+        {
+            return Sse.SetVector128(src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);
+        }
+
+        private static Vector128<float> Rotate(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x39);
+        }
+
+        private static Vector128<float> RotateReverse(Vector128<float> x)
+        {
+            return Sse.Shuffle(x, x, 0x93);
+        }
+
+        // Warning: this operation changes the value of x => do not reuse x
+        private static unsafe void Store4(Vector128<float> x, float* dst, int* idx)
+        {
+            Sse.StoreScalar(dst + idx[0], x);
+            for (int i = 1; i <= 3; i++)
+            {
+                x = Rotate(x);
+                Sse.StoreScalar(dst + idx[i], x);
+            }
+        }
+
+        internal static void ScaleU(float scale, float[] dst, int count)
+        {
+            Vector128<float> scaleVector = Sse.SetAllVector128(scale);
+
+            unsafe
+            {
+                fixed (float* pdst = dst)
+                {
+                    float* pDstCurrent = pdst;
+                    float* pEnd = pdst + count;
+
+                    while (pDstCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
```

Comment:
@tannergooding - do you think it is worthwhile to do the "preamble", like we do in https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs#L96-L102?

@briancylui - this comment wouldn't need to be addressed in the first PR, but if necessary, we could address it in a future revision.

Reply:
I think it would be useful to link to the software implementation with a comment indicating that this a vectorized version of the software/serialized version.

Reply:
What I was specifically asking about is getting aligned at the beginning of the array - that's the "preamble" part I was referring to.

1. while (!aligned) { do scalar operation; }  // preamble
2. Do vectorized operation using Read**Aligned**
3. while (!end) { do scalar operation; }

Reply:
Ah, I see. Yes, we should. For large arrays, especially those that cross cache line or page boundaries, doing this should save some measurable amount of time.